### PR TITLE
SimTelFile

### PR DESCRIPTION
**Pull request by dneise**
I branched here the `SimTelFile` off of the current varint_cython branch.

This time I played with the older `gamma_test` file and not with the newer prod4 files .. turns out their structures are slightly different: There can be multiple of these in the header before the actual showers start:
 `[History, SimTelMCRunHeader, CORSIKAInputCard ]`

I did not yet research what it means, if there are multiple of these, I just wanted to see if  the current `SimTelFile` finds the same number of events as the `HESSIOEventSource`, c.f. #83 

I wrote this little try-script:
```
#!/usr/bin/env python
import sys
import time
from eventio.simtel.simtelfile import SimTelFile

url = sys.argv[1]
source = SimTelFile(url)

start_time = time.time()
for i, e in enumerate(source):
    print(e)
    pass
print('  Duration:', time.time() - start_time)
print("i:", i)
```

Result looks like this:
```
11:16 $ ./try_simtelfile.py eventio/resources/gamma_test.simtel.gz 
(SimTelMCShower[2020](size=620, only_subobjects=False, first_byte=29881604), {'mc_event': SimTelMCEvent[2021](size=12, only_subobjects=False, first_byte=29882464), 'pe_sum': SimTelMCPeSum[2026](size=3536, only_subobjects=False, first_byte=36168524), 'event': SimTelEvent[2010](glob_count=408, size=116608, first_byte=36172076)})
(SimTelMCShower[2020](size=620, only_subobjects=False, first_byte=29881604), {'mc_event': SimTelMCEvent[2021](size=12, only_subobjects=False, first_byte=36288700), 'pe_sum': SimTelMCPeSum[2026](size=3536, only_subobjects=False, first_byte=46565696), 'event': SimTelEvent[2010](glob_count=409, size=1122888, first_byte=46569248)})
(SimTelMCShower[2020](size=620, only_subobjects=False, first_byte=47694900), {'mc_event': SimTelMCEvent[2021](size=12, only_subobjects=False, first_byte=47695620), 'pe_sum': SimTelMCPeSum[2026](size=3536, only_subobjects=False, first_byte=47695648), 'event': SimTelEvent[2010](glob_count=803, size=796152, first_byte=47699200)})
(SimTelMCShower[2020](size=620, only_subobjects=False, first_byte=48546384), {'mc_event': SimTelMCEvent[2021](size=12, only_subobjects=False, first_byte=48547216), 'pe_sum': SimTelMCPeSum[2026](size=3536, only_subobjects=False, first_byte=48547244), 'event': SimTelEvent[2010](glob_count=4907, size=471780, first_byte=48550796)})
(SimTelMCShower[2020](size=620, only_subobjects=False, first_byte=49078076), {'mc_event': SimTelMCEvent[2021](size=12, only_subobjects=False, first_byte=49082488), 'pe_sum': SimTelMCPeSum[2026](size=3536, only_subobjects=False, first_byte=49082516), 'event': SimTelEvent[2010](glob_count=9508, size=403972, first_byte=49086068)})
(SimTelMCShower[2020](size=620, only_subobjects=False, first_byte=49494664), {'mc_event': SimTelMCEvent[2021](size=12, only_subobjects=False, first_byte=49495412), 'pe_sum': SimTelMCPeSum[2026](size=3536, only_subobjects=False, first_byte=49495440), 'event': SimTelEvent[2010](glob_count=10104, size=138564, first_byte=49498992)})
(SimTelMCShower[2020](size=620, only_subobjects=False, first_byte=49494664), {'mc_event': SimTelMCEvent[2021](size=12, only_subobjects=False, first_byte=49637684), 'pe_sum': SimTelMCPeSum[2026](size=3536, only_subobjects=False, first_byte=49637712), 'event': SimTelEvent[2010](glob_count=10109, size=137592, first_byte=49641264)})
(SimTelMCShower[2020](size=620, only_subobjects=False, first_byte=49801548), {'mc_event': SimTelMCEvent[2021](size=12, only_subobjects=False, first_byte=49802324), 'pe_sum': SimTelMCPeSum[2026](size=3536, only_subobjects=False, first_byte=49802352), 'event': SimTelEvent[2010](glob_count=11905, size=195732, first_byte=49805904)})
(SimTelMCShower[2020](size=620, only_subobjects=False, first_byte=50003596), {'mc_event': SimTelMCEvent[2021](size=12, only_subobjects=False, first_byte=50004288), 'pe_sum': SimTelMCPeSum[2026](size=3536, only_subobjects=False, first_byte=50004316), 'event': SimTelEvent[2010](glob_count=12202, size=1782080, first_byte=50007868)})
(SimTelMCShower[2020](size=620, only_subobjects=False, first_byte=50003596), {'mc_event': SimTelMCEvent[2021](size=12, only_subobjects=False, first_byte=51789964), 'pe_sum': SimTelMCPeSum[2026](size=3536, only_subobjects=False, first_byte=51789992), 'event': SimTelEvent[2010](glob_count=12203, size=1560656, first_byte=51793544)})
  Duration: 0.30574679374694824
i: 9
```

### CODE REVIEW

**Comment by dneise** [comment]
I think we should work on the `__repr__`s. This is not really interesting anymore, I think

    size=12, only_subobjects=False, first_byte=29882464

now I am mostly interested in  `self.header.id` and such.

**Comment by dneise** [comment]
Such a testfile:
```python
#!/usr/bin/env python
import sys
import time
from pprint import pprint
from eventio.simtel.simtelfile import SimTelFile

url = sys.argv[1]
source = SimTelFile(url)

start_time = time.time()
for i, e in enumerate(source):
    print('Event id:', i)
    pprint(e)
    pass
print('  Duration:', time.time() - start_time)
```

Creates this output:

----

```
12:15 $ ./try_simtelfile.py eventio/resources/gamma_test.simtel.gz
Event id: 0
(SimTelMCShower[2020](size=620, only_subobjects=False, first_byte=29881604),
 {'event': {'cent_event': SimTelCentEvent[2009](size=48, only_subobjects=False, first_byte=12),
            'tel_events': [SimTelTelEvent[2238](telescope_id=38, size=58572, first_byte=72),
                           SimTelTelEvent[2247](telescope_id=47, size=57912, first_byte=58656)],
            'track_events': [SimTelTrackEvent[2138](telescope_id=38, size=8, first_byte=116580),
                             SimTelTrackEvent[2147](telescope_id=47, size=8, first_byte=116600)]},
  'mc_event': [SimTelMCEvent[2021](size=12, only_subobjects=False, first_byte=29882240),
               SimTelMCEvent[2021](size=12, only_subobjects=False, first_byte=29882268),
               SimTelMCEvent[2021](size=12, only_subobjects=False, first_byte=29882296),
               SimTelMCEvent[2021](size=12, only_subobjects=False, first_byte=29882324),
               SimTelMCEvent[2021](size=12, only_subobjects=False, first_byte=29882352),
               SimTelMCEvent[2021](size=12, only_subobjects=False, first_byte=29882380),
               SimTelMCEvent[2021](size=12, only_subobjects=False, first_byte=29882408),
               SimTelMCEvent[2021](size=12, only_subobjects=False, first_byte=29882436),
               SimTelMCEvent[2021](size=12, only_subobjects=False, first_byte=29882464)],
  'pe_sum': SimTelMCPeSum[2026](size=3536, only_subobjects=False, first_byte=36168524)})

....
```

**Comment by dneise** [comment]
Okay it seems there is no seeking backwards at the moment ...

**Comment by dneise** [comment]
try it out like:
```
21:03 $ ./examples/try_simtelfile.py \
eventio/resources/gamma_20deg_0deg_run102___cta-prod4-sst-1m_desert-2150m-Paranal-sst-1m.simtel.gz 
Event id: 0
Event id: 1
Event id: 29
Event id: 30
  Duration: 0.607128381729126

21:04 $ ./examples/try_simtelfile.py \
eventio/resources/gamma_20deg_0deg_run102___cta-prod4-sst-1m_desert-2150m-Paranal-sst-1m.simtel.zst 
Event id: 0
Event id: 1
...
Event id: 29
Event id: 30
  Duration: 0.49201059341430664
```

**Comment by dneise** [comment]
If you wanna look at the events returned by `SimTelFile` use `try_simtelfile.py -p`

File: eventio/simtel/simtelfile.py [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+import logging
+from eventio.base import EventIOFile, EventIOObject
+
+class WrongType(Exception):
+    pass
+
+class NoTrackEvents(Exception):
+    pass
+
+class WithNextAssert:
```
Comment (MaxNoe) [inline eventio/simtel/simtelfile.py]: I think classes having this mixed in should also use `_last_obj` in the plain `__next__`, so users don't shoot themselvees in the foot...

File: eventio/simtel/simtelfile.py [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+
+        o = self._last_obj
+        if not isinstance(o, object_):
+            return None
+
+        self._last_obj = None
+        return o
+
+
+EventIOObject.next_assert = WithNextAssert.next_assert
```
Comment (MaxNoe) [inline eventio/simtel/simtelfile.py]: this looks like a hack. Why not put it in base?

File: eventio/simtel/simtelfile.py [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+                ).parse_data_field()
+                self.lascal[lascal['telescope_id']] = lascal
+
+        except WrongType:
+            # this is normal .. not every event has updates here
+            pass
+
+
+
+class EventIOFileWithNextAssert(EventIOFile, WithNextAssert):
```
Comment (MaxNoe) [inline eventio/simtel/simtelfile.py]: same here, does this new functionality have a large overhead? Or could we just put this in `base.py`?

File: eventio/simtel/simtelfile.py [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+    pass
+
+
+def telescope_description_from(file_):
```
Comment (MaxNoe) [inline eventio/simtel/simtelfile.py]: add `get_` or `build_`? So the function is a verb?

File: eventio/simtel/simtelfile.py [hunk 5]
```diff
@@ -0,0 +1,394 @@
...
+        file_.next_assert(SimTelCamOrgan),
+        file_.next_assert(SimTelPixelset).parse_data_field(),
+        file_.next_assert(SimTelPixelDisable),
+        file_.next_assert(SimTelCamsoftset),
+        file_.next_assert(SimTelTrackSet),
+        file_.next_assert(SimTelPointingCor),
+    ]
+
+
+def read_all_of_type(f, type_, converter=lambda x: x):
```
Comment (MaxNoe) [inline eventio/simtel/simtelfile.py]: is this useful in a broader context? So maybe put it into `search_utils.py`?

File: eventio/simtel/simtelfile.py (hunk 4 above)
Comment (dneise) [reply eventio/simtel/simtelfile.py]: never get .. get adds no meaning just noise .. if have `a.mean()` or `a.get_mean()` .. what do you prefer?

Comment (dneise) [reply eventio/simtel/simtelfile.py]: let me see .. is it fetch or read maybe?

Comment (MaxNoe) [reply eventio/simtel/simtelfile.py]: Just Peace it like it is

**Pull request merged**